walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -134,56 +156,179 @@
+      int availableTimeRangeEnd = startOrderedTimeRanges[startOrderedPointer].start();
       checkDurationAndAddAvailableTimeRange(requestedDuration,
                                             availableTimeRangeStart, availableTimeRangeEnd,
                                             availableTimeRanges);
 
       if (startOrderedPointer + 1 < startOrderedTimeRanges.length &&
           endOrderedPointer + 1 < endOrderedTimeRanges.length) {
-        startingTimeRange = startOrderedTimeRanges[++startOrderedPointer];
-        endingTimeRange = endOrderedTimeRanges[++endOrderedPointer];
+        startOrderedPointer++;
+        endOrderedPointer++;
       } else {
         return;
       }
     }
   }
 
+  /**
+   * Finds the overlapping, available {@code TimeRange} for mandatory and optional attendees. 
+   * If no {@code TimeRange} satisfies all mandatory and optional attendees, returns those
+   * {@code TimeRange} that satisfies mandatory attendees.
+   *
+   * @param requestedDuration The requested length of the overlapping, available {@code TimeRange}
+   *     to find.
+   * @param availableTimeRangesForMandatoryAttendees A list of available {@code TimeRange} for the
+   *     mandatory attendees. This list is expected to be sorted in ascending chronological order.
+   * @param availableTimeRangesForOptionalAttendees A list of available {@code TimeRange} for the
+   *     optional attendees. This list is expected to be sorted in ascending chronological order.
+   * @return A list of all available {@code TimeRange} for both mandatory and optional attendees,
+   *     sorted in ascending chronological order.
+   */
+  private List<TimeRange> getOverlappingAvailabeTimeRanges(long requestedDuration,
+      List<TimeRange> availableTimeRangesForMandatoryAttendees,
+      List<TimeRange> availableTimeRangesForOptionalAttendees) {
+    if (availableTimeRangesForMandatoryAttendees.isEmpty()) {
+      return Arrays.asList();
+    } else if (availableTimeRangesForOptionalAttendees.isEmpty()) {
+      return availableTimeRangesForMandatoryAttendees;
+    }
+
+    List<TimeRange> availableTimeRangesForMandatoryOptionalAttendees = new LinkedList<>();
+    findAndAddOverlappingAvailableTimeRanges(requestedDuration,
+                                             availableTimeRangesForMandatoryAttendees,
+                                             availableTimeRangesForOptionalAttendees,
+                                             availableTimeRangesForMandatoryOptionalAttendees);
+    return availableTimeRangesForMandatoryOptionalAttendees;
+  }
+
+  /**
+   * Finds and adds the overlapping {@code TimeRange} for mandatory and optional attendees. These
+   * {@code TimeRange} have at least a length of the requested duration.
+   */
+  private void findAndAddOverlappingAvailableTimeRanges(long requestedDuration,
+      List<TimeRange> timeRangesForMandatoryAttendeesList,
```

Comment:
The variable names are getting a bit too verbose. Consider renaming to
mandatoryTimeRanges
optionalTimeRanges etc.